### PR TITLE
Tooling to install and start cassandra clusters

### PR DESCRIPTION
**Pull request by petermattis**

### CODE REVIEW

File: cassandra.go [hunk 1]
```diff
@@ -0,0 +1,1327 @@
...
+
+func (cassandra) start(c *cluster) {
+	yamlPath, err := makeCassandraYAML(c)
+	if err != nil {
+		log.Fatal(err)
+	}
+	c.put(yamlPath, "./cassandra.yaml")
+	_ = os.Remove(yamlPath)
+
+	display := fmt.Sprintf("%s: starting cassandra (be patient)", c.name)
```
Comment (tschottdorf) [inline cassandra.go]: > (be patient)

🙈

File: cassandra.go [hunk 2]
```diff
@@ -0,0 +1,1327 @@
...
+	}
+	if err := t.Execute(bufio.NewWriter(f), m); err != nil {
+		log.Fatal(err)
+	}
+	return f.Name(), nil
+}
+
+const cassandraYAML = `# Cassandra storage config YAML
+
+# The name of the cluster. This is mainly used to prevent machines in
```
Comment (tschottdorf) [inline cassandra.go]: This is such a pile of text that you might as well put it into a `cassandra_yaml.go` (I understand that you want it compiled in which makes sense).

Comment (tschottdorf) [reply cassandra.go]: I assume this is the default yaml file. Perhaps you can even make that more explicit and concatenate any of our updates to it (simulating a "diff")? I assume latest seen takes precedence in this file.

If you don't think we're ever going to change this file, this doesn't seem necessary.

File: cockroach.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+type cockroach struct{}
+
+func (r cockroach) start(c *cluster) {
+	display := fmt.Sprintf("%s: starting", c.name)
+	host1 := c.host(1)
+	nodes := c.serverNodes()
+	bootstrapped := false
+	c.parallel(display, len(nodes), 0, func(i int) ([]byte, error) {
+		if nodes[i] == 1 {
+			bootstrapped = true
```
Comment (tschottdorf) [inline cockroach.go]: I guess this isn't technically a data race, but it looks fishy enough to just use an atomic.

File: cluster.go [hunk 4]
```diff
@@ -47,75 +53,29 @@ func (c *cluster) cockroachNodes() []int {
...
-		args = append(args, "--insecure")
-	}
-	args = append(args, "--store=path=/mnt/data1")
-	args = append(args, "--logtostderr")
-	args = append(args, "--log-dir=")
-	args = append(args, "--background")
-	args = append(args, "--cache=50%")
-	if join != host {
-		args = append(args, "--join="+join)
+	cmd := `ip addr | grep 'state UP' -A2 | tail -n1 | awk '{print $2}' | cut -f1  -d'/'`
```
Comment (tschottdorf) [inline cluster.go]: You can't just use `hostname --all-ip-addresses` which seems harder to screw up the processing of? Or something else, like the network interface's name.

File: cassandra.go (hunk 1 above)
Comment (petermattis) [reply cassandra.go]: Yeah, it takes 1m15s to start a 4-node cluster. Each node takes ~20s to start and a node refuses to join the cluster when another node is joining. At least, this is what I've found experimentally. @mberhault do you have any further knowledge to add here?

File: cockroach.go (hunk 3 above)
Comment (petermattis) [reply cockroach.go]: I can also determine if node 1 was started outside the loop.

File: cluster.go (hunk 4 above)
Comment (petermattis) [reply cluster.go]: I grabbed this shell pipeline from stackoverflow. You're clearly smarter than the interwebs. Changed.

File: cassandra.go (hunk 2 above)
Comment (petermattis) [reply cassandra.go]: Hmm, that's a good idea. Let me see if it works.

Comment (petermattis) [reply cassandra.go]: It mostly works. I had to comment out `listen_address` because I couldn't figure out how to clear it once it was set.

**Pull request merged**